subprojects/core/src/main/java/org/gradle/cache/internal/DefaultCrossBuildInMemoryCacheFactory.java

```diff
@@ -232,4 +242,52 @@ protected V maybeGetRetainedValue(Class<?> key) {
             return leakyValues;
         }
     }
+
+    private static class BuildAgeEvictionCrossBuildInMemoryCache<K, V> implements CrossBuildInMemoryCache<K, V>, SessionLifecycleListener {
+        private final ManualEvictionInMemoryCache<K, V> delegate = new ManualEvictionInMemoryCache<>();
+        private final ConcurrentMap<K, Boolean> keysFromPreviousBuild = new ConcurrentHashMap<>();
+        private final ConcurrentMap<K, Boolean> keysFromCurrentBuild = new ConcurrentHashMap<>();
+
+        @Override
+        public V get(K key, Function<? super K, ? extends V> factory) {
+            keysFromCurrentBuild.put(key, Boolean.TRUE);
+            return delegate.get(key, factory);
+        }
+
+        @Override
+        public V get(K key) {
+            keysFromCurrentBuild.put(key, Boolean.TRUE);
+            return delegate.get(key);
+        }
+
+        @Override
+        public void put(K key, V value) {
+            keysFromCurrentBuild.put(key, Boolean.TRUE);
+            delegate.put(key, value);
+        }
+
+        @Override
+        public void clear() {
+            delegate.clear();
+            keysFromCurrentBuild.clear();
+            keysFromPreviousBuild.clear();
+        }
+
+        @Override
+        public void afterStart() {
+        }
+
+        @Override
+        public void beforeComplete() {
+            final Set<K> keysToRetain = new HashSet<>();
+            keysToRetain.addAll(keysFromPreviousBuild.keySet());
+            keysToRetain.addAll(keysFromCurrentBuild.keySet());
+
+            delegate.retainAll(keysToRetain);
+
+            keysFromPreviousBuild.clear();
+            keysFromPreviousBuild.putAll(keysFromCurrentBuild);
+            keysFromCurrentBuild.clear();
+        }
+    }
```

Comment:
I don't see a test case for this where we actually test the eviction, only that some things are retained when they should be. I think we should add such a test.

Reply:
I added a test case for the eviction.